### PR TITLE
Narrow constraints to `IsRecentEra era` in `Shelley.Transaction` module

### PR DESCRIPTION
**Pull request by Anviking**
- Replace `IsShelleyBasedEra` and `IsCardanoEra` and similar with `IsRecentEra` / `RecentEra era`
- [x] Error appropriately from signTransaction with non-recent tx formats

### Issue Number

ADP-2353 — increase consistency first

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 1]
```diff
@@ -600,32 +573,15 @@ newTransactionLayer keyF networkId = TransactionLayer
...
-                    & sealedTxFromCardano'
+                InAnyCardanoEra txEra (Cardano.Tx body wits) ->
+                    case Write.toRecentEra txEra of
+                        Just recentTxEra -> Write.withConstraints recentTxEra $
+                            signTransaction
+                                keyF networkId witCountCtx acctResolver
+                                policyResolver policyKeyM stakingScriptResolver
+                                addressResolver inputResolver (body, wits)
+                            & sealedTxFromCardano'
+                        Nothing -> sealedTx
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: Should probably error... It matters more now that we no longer support signing  `Shelley, ..., Alonzo` txs

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs [hunk 2]
```diff
@@ -357,7 +357,9 @@ type RecentEraLedgerConstraints era =
 -- https://cardanofoundation.atlassian.net/browse/ADP-2353
 withConstraints
     :: RecentEra era
-    -> ((RecentEraLedgerConstraints (CardanoApi.ShelleyLedgerEra era)) => a)
+    -> ( ( IsRecentEra era
+         , (RecentEraLedgerConstraints (CardanoApi.ShelleyLedgerEra era))
+         ) => a)
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs]: At first glance this seems a bit awkward to read (and to format).

Why not define a separate constraint synonym:
```hs
type RecentEraConstraints era =
    ( IsRecentEra era
    , RecentEraLedgerConstraints (CardanoApi.ShelleyLedgerEra era)
    )
```

Then we can write:
```hs
withConstraints :: RecentEra era -> (RecentEraConstraints era => a) -> a
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @Anviking 

Would love to discuss this during our call later, so I can better understand the motivation and direction!

Nevertheless, I've put some suggestions in https://github.com/cardano-foundation/cardano-wallet/pull/4215. These suggestions are mainly fixes to formatting and documentation.

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]: See: 8f70f9bf73ae5c8beb3261ea917818130ea8ba09

**Review by jonathanknowles — APPROVED** [review]

**Pull request merged**